controllers/topology/internal/extensions/patches/engine.go

```diff
@@ -0,0 +1,494 @@
+	}
+	t.Variables = controlPlaneVariables
+	req.Items = append(req.Items, t)
+
+	// If the clusterClass mandates the controlPlane has infrastructureMachines,
+	// add the InfrastructureMachineTemplate for control plane machines.
+	if blueprint.HasControlPlaneInfrastructureMachine() {
+		t, err = newTemplate(blueprint.ControlPlane.InfrastructureMachineTemplate, api.TemplateID{
+			APIVersion: blueprint.ControlPlane.InfrastructureMachineTemplate.GetAPIVersion(),
+			Kind:       blueprint.ControlPlane.InfrastructureMachineTemplate.GetKind(),
+			TargetType: api.ControlPlaneInfrastructureMachineTemplateTargetType,
+		})
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create ControlPlane's machine template %s",
+				tlog.KObj{Obj: blueprint.ControlPlane.InfrastructureMachineTemplate})
+		}
+		t.Variables = controlPlaneVariables
+		req.Items = append(req.Items, t)
+	}
+
+	// Add BootstrapConfigTemplate and InfrastructureMachine template for all MachineDeploymentTopologies
+	// in the Cluster.
+	// NOTE: We intentionally iterate over MachineDeploymentTopologies in the Cluster instead of over
+	// MachineDeploymentClasses in the ClusterClass. We have to calculate patches per MachineDeploymentTopology
+	// instead of per MachineDeploymentClass, because the builtin variable values depend on the current
+	// MachineDeployment. E.g., there are variables for the version or replicas of the MachineDeployment.
+	for mdTopologyName, md := range desired.MachineDeployments {
+		// Lookup MachineDeploymentTopology.
+		mdTopology, err := lookupMDTopology(blueprint.Topology, mdTopologyName)
+		if err != nil {
+			return nil, err
+		}
+
+		// Get corresponding MachineDeploymentClass.
+		mdClass, ok := blueprint.MachineDeployments[mdTopology.Class]
+		if !ok {
+			return nil, errors.Errorf("failed to lookup MachineDeployment class %q in ClusterClass", mdTopology.Class)
+		}
+
+		// Calculate MachineDeployment variables.
+		mdVariables, err := variables.MachineDeployment(mdTopology, md.Object)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to calculate variables for %s", tlog.KObj{Obj: md.Object})
+		}
+
+		// Add the BootstrapTemplate.
+		t, err := newTemplate(mdClass.BootstrapTemplate, api.TemplateID{
+			APIVersion: mdClass.BootstrapTemplate.GetAPIVersion(),
+			Kind:       mdClass.BootstrapTemplate.GetKind(),
+			TargetType: api.MachineDeploymentBootstrapConfigTemplateTargetType,
+			MachineDeployment: api.MachineDeploymentID{
+				TopologyName: mdTopologyName,
+				Class:        mdTopology.Class,
+			},
+		})
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create BootstrapConfig template %s for MachineDeployment topology %s",
+				tlog.KObj{Obj: mdClass.BootstrapTemplate}, mdTopologyName)
+		}
+		t.Variables = mdVariables
+		req.Items = append(req.Items, t)
+
+		// Add the InfrastructureMachineTemplate.
+		t, err = newTemplate(mdClass.InfrastructureMachineTemplate, api.TemplateID{
```

Comment:
All of the `api.TemplateID` seems contain lots of repetition. The APIVersion/Kind could be reused, while the target type can be passed into a builder, same with the MachineDeployment